compiler/src/iree/compiler/Codegen/LLVMCPU/EncodingInfo.cpp

```diff
@@ -0,0 +1,193 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/LLVMCPU/EncodingInfo.h"
+
+#include "iree-dialects/Dialect/LinalgExt/IR/LinalgExtDialect.h"
+#include "iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.h"
+#include "iree-dialects/Dialect/LinalgExt/Passes/Passes.h"
+#include "iree-dialects/Dialect/LinalgExt/Utils/Utils.h"
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/HAL/IR/HALDialect.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "mlir/Dialect/Arith/Utils/Utils.h"
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace {
+
+/// Get the materialization information from a `iree_linalg_ext.pack` operation.
+static FailureOr<IREE::LinalgExt::MaterializeEncodingInfo>
+getMaterializationInfo(IREE::LinalgExt::PackOp packOp) {
+  IREE::LinalgExt::MaterializeEncodingInfo encodingInfo;
+  SmallVector<OpFoldResult> mixedTileSizes = packOp.getMixedTiles();
+  encodingInfo.innerTileSizes.reserve(mixedTileSizes.size());
+  for (auto tileSize : mixedTileSizes) {
+    if (tileSize.is<Value>()) {
+      encodingInfo.innerTileSizes.push_back(ShapedType::kDynamic);
+    } else {
+      encodingInfo.innerTileSizes.push_back(
+          tileSize.get<Attribute>().cast<IntegerAttr>().getInt());
+    }
+  }
+  encodingInfo.innerDimsPos = llvm::to_vector(packOp.getInnerDimsPos());
+  encodingInfo.outerDimsPerm = llvm::to_vector(packOp.getOuterDimsPerm());
+  encodingInfo.srcRank = packOp.getInputRank();
+  return encodingInfo;
+}
+
+/// Pattern to lower a `flow.dispatch.workgroup_count_from_set_encoding` op.
+/// At the Flow level this op uses the logical shape of the tensor
+/// as the workload. This gets materialized into an physical tensor
+/// Lower this operation accounting for the change of shape from
+/// the logical shape to the physical shape. It lowers to
+/// a `flow.dispatch.workgroup_count_from_root_dag` where the root
+/// is the `pack` op that materialized the encoding.
+struct LowerDispatchWorkgroupCountFromSetEncodingOp
```

Comment:
Could we keep this pattern where it is. I think as long as the `encodingInfo` and `materializeEncodingValueFn`, etc. are provided, this is the same on all backends.....
All this is tech debt and should go away anyway.

Reply:
I thought it is related to how we materialize the encoding. E.g. if the set_encoding op is converted to a pack op, then the iteration domain is on packed layout. If the set encoding op is converted to a pad op, then the iteration domain is in a padded domain. The pattern is specifically for pack ops?

Reply:
After thinking a bit more, I think you're right. This is the same on all the backends because the pack op is already generated.